chalice/compat.py

```diff
@@ -1,8 +1,112 @@
+
+    # This is the setuptools shim used to execute setup.py by pip.
+    # Lines 2 and 3 have been added to call the above function
+    # `prevent_msvc_compiling_patch` and extra escapes have been added on line
+    # 5 because it is passed through another layer of string parsing before it
+    # is executed.
+    _SETUPTOOLS_SHIM = (
+        "import setuptools, tokenize;__file__=%r;"
+        "from chalice.compat import prevent_msvc_compiling_patch;"
+        "prevent_msvc_compiling_patch();"
+        "f=getattr(tokenize, 'open', open)(__file__);"
+        "code=f.read().replace('\\\\r\\\\n', '\\\\n');"
+        "f.close();"
+        "exec(compile(code, __file__, 'exec'))"
+    )
+
+    # On windows the C compiling story is much more complex than on posix as
+    # there are many different C compilers that setuptools and disutils will
+    # try and find using a combination of known filepaths, registry entries,
+    # and environment variables. Since there is no simple environment variable
+    # we can replace when starting the subprocess that builds the package;
+    # we need to apply a patch at runtime to prevent pip/setuptools/distutils
+    # from being able to build C extensions.
+    # Patching out every possible technique for finding each compiler would
+    # be a losing game of whack-a-mole. In addition we need to apply a patch
+    # two layers down through subprocess calls, specifically:
+    #  * Chalice creates a subprocess of `pip wheel ...` to build sdists
+    #    into wheel files.
+    #  * Pip creates another python subprocess to call the setup.py file in
+    #    the sdist. Before doing so it applies the above shim to make the
+    #    setup file compatible with setuptools. This shim layer also reads
+    #    and executes the code in the setup.py.
+    #  * Setuptools (which will have been executed by the shim) will
+    #    eventually call distutils to do the heavy lifting for C compiling.
+    #
+    # Our patch needs to affect the bottom level here (distutils) and patch
+    # it out to prevent it from compiling C in a graceful way that results in
+    # falling back to building a purepython library if possible.
+    # The below line will be injected just before the `pip wheel ...` portion
+    # of the subprocess that Chalice starts. This replaces the
+    # SETUPTOOLS_SHIM that pip normally uses with the one defined above.
+    # When pip goes to run its subprocess for executing setup.py it will
+    # inject _SETUPTOOLS_SHIM rather than the usual SETUPTOOLS_SHIM in pip.
+    # This lets us apply our patches in the same process that will compile
+    # the c extensions before the setup.py file has been executed.
+    # The actual patches used are decribed in the comment above
+    # _SETUPTOOLS_SHIM.
+    pip_no_compile_c_shim = (
+        "import pip;"
+        "pip.wheel.SETUPTOOLS_SHIM = \"\"\"%s\"\"\";"
+    ) % _SETUPTOOLS_SHIM
+    pip_no_compile_c_env_vars = {}  # type: ignore
```

Comment:
I'd still like to get these typed.  These are passed to code that we own in package.py.